### PR TITLE
Add SpeedTree8 shadergraph to HD and Universal

### PR DESCRIPTION
**Pull request by nerites**
# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [x] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Adding SpeedTree8 support (including material upgrader) to HDRP.
Added a graphics test in HDRP for SpeedTree8. Also added SpeedTrees to the terrainlit and terrainbaselit test scenes.
Added the SpeedTree8 shadergraph to URP and replacing existing shader as default.
Replaced the SpeedTree8 shaders used in universal tests that involve ST8 assets.

TODO: Move SpeedTree8MaterialUpgrader from Core to Shadergraph package? -> Even invoking MaterialUpgrader using Core.Editor.MaterialUpgrader does not find the class from the shadergraph package, for some reason. Why?

21.1 backport: https://github.com/Unity-Technologies/Graphics/pull/4068
10.x.x backport: https://github.com/Unity-Technologies/Graphics/pull/3861/

---
### Testing status
Added a SpeedTree8 test to HDRP and added ST8 assets to the TerrainLit + Basemap tests in HDRP. 

Tested importing new ST8 assets in both Universal and HD. 
Also tested upgrading, but that was many changes ago and is worth re-testing. To do this, must comment out the OnPostprocessSpeedTreeAsset callback in the MaterialPostprocessor in either pipeline, then add a ST8 asset to the project. Then, running the upgrader should preserve properties such as wind quality, billboard, etc.

Tested Wind, Subsurface, and billboard work correctly on both.

Need to test huevariation.

Backface culling is not working on the Universal shadergraph- I need to set up some way of setting that per-material in Universal. I'm thinking of just copying what the HD Lit target does. 

Universal Shadergraph tests look significantly different in Yamato runs of tests, even though running locally passes just fine. This needs investigation.

---

### CODE REVIEW

**Comment by nerites** [comment]
Backport is here: https://github.com/Unity-Technologies/Graphics/pull/3861

**Comment by nerites** [comment]
Backport to 2021.1: https://github.com/Unity-Technologies/Graphics/pull/4068

**Review by LandonTownsendUnity — APPROVED** [review]
I think I've tested this enough, it should be good now.

**Comment by nerites** [comment]
The remaining Yamato failures are all hybrid tests that timeout due to either a firewall or nvidia desktop popup, or pipelines that have a dependency on a failing hybrid test. These failures occur in master as well.

**Review by elizabeth-legros — APPROVED** [review]

**Review by xiaoxicici — APPROVED** [review]
Tested the latest in HDRP, notes can be found in this doc: https://docs.google.com/document/d/1vue-NcZoeNGGHMlLTRiqWY-QcXy0Hx4Tv6IWXn4ioZw/edit#

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl [hunk 1]
```diff
@@ -91,6 +91,15 @@ struct SpeedTreeFragmentInput
...
 
+#define UNITY_INV_PI        0.31830988618f
+inline float GGXTerm (float NdotH, float roughness)
+{
+    float a2 = roughness * roughness;
+    float d = (NdotH * a2 - NdotH) * NdotH + 1.0f; // 2 mad
+    return UNITY_INV_PI * a2 / (d * d + 1e-7f); // This function is not intended to be running on Mobile,
+                                            // therefore epsilon is smaller than what can be represented by half
+}
+
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl:102]: In BSDF.hlsl we seem to have the same functions for this calculations. In order not to duplicate things, I would recommend trying to use those.
Check the `real D_GGXNoPI(real NdotH, real roughness)` in that file.

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader [hunk 2]
```diff
@@ -66,7 +66,7 @@ Shader "Universal Render Pipeline/Nature/SpeedTree8"
             #pragma shader_feature_local _WINDQUALITY_NONE _WINDQUALITY_FASTEST _WINDQUALITY_FAST _WINDQUALITY_BETTER _WINDQUALITY_BEST _WINDQUALITY_PALM
             #pragma shader_feature_local EFFECT_BILLBOARD
             #pragma shader_feature_local EFFECT_HUE_VARIATION
-            //#pragma shader_feature_local EFFECT_SUBSURFACE // GI dependent.
+            #pragma shader_feature_local EFFECT_SUBSURFACE
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader:69]: Could this be made `shader_feature_local_fragment` ?

Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader]: Changed in d2a779d

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl (hunk 1 above)
Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8Passes.hlsl]: Changed in d2a779d

File: com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader (hunk 2 above)
Comment (nerites) [reply com.unity.render-pipelines.universal/Shaders/Nature/SpeedTree8.shader]: After testing on yamato, changing to shader_feature_local_fragment breaks iphone tests. Reverting.

**Review by ValGrimm-U3D — APPROVED** [review]
Thanks for responding to my many questions and suggestions, fine to merge.  👍

**Review by JMargevics — APPROVED** [review]
[Test doc](https://confluence.unity3d.com/pages/viewpage.action?pageId=129564990)

❕ One major issue left: no trees visible in Metal build. @nerites is aware of this issue, it will take time to get fixed. We don't want to stall this PR any further, so I'm approving this as is. Will report this to FogBugz for better tracking.

We will be working to add ST8+terrains coverage into runtime tests 👍

✅ Otherwise there are no major issue on all other platforms. PR has been extensively tested, including consoles. ST8 shader and importer work nicely with wind physics and HDRP systems.

**Review by ellioman — APPROVED** [review]

**Comment by nerites** [comment]
The remaining failure in URPUpdate_Top_Asset_Store is fixed in this PR and not due to my changes: https://github.cds.internal.unity3d.com/sophia/URP-Update-testing/pull/1

**Comment by lumonix** [comment]
Grant and I have discussed that we will allow this PR to merge even though URP QA has not been able to finish their testing on this, because we do not want to miss the 10.5 deadline.
We've done URP testing on our end and will follow up promptly with any bug fixes in case URP QA finds any.

**Comment by GrantLamb-Unity** [comment]
Update after further discussions - Since it looks like the deadline is by EoD tomorrow ( 4/21 ) I'd like for us to not merge and I'll chat with URP QA. They should be the ones who would allow a merge at this point or not.

Additionally Landon is out today and I'd like to ask him some follow up questions about his tests, which we could potentially share with URP and see if it's possible that we have sufficient coverage via the tests that have already been done.

**Pull request merged**